src/bootstrap/builder.rs

```diff
@@ -184,6 +184,25 @@ impl StepDescription {
         let should_runs =
             v.iter().map(|desc| (desc.should_run)(ShouldRun::new(builder))).collect::<Vec<_>>();
 
+        fn match_run(
+            v: &[StepDescription],
+            builder: &Builder<'_>,
+            should_runs: &[ShouldRun<'_>],
+            path: &Path,
+        ) -> bool {
+            let mut attempted_run = false;
+            for (desc, should_run) in v.iter().zip(should_runs) {
```

Comment:
Could you refactor this (ideally in a preceding separate commit) to make should_runs contain (StepDescription, ShouldRun) tuples rather than needing to zip the two together in each iteration? It looks like that should be pretty easy and would simplify the signatures a bit here.